game-core/src/main/java/games/strategy/engine/history/change/units/KillUnits.java

```diff
@@ -0,0 +1,125 @@
+package games.strategy.engine.history.change.units;
+
+import games.strategy.engine.data.CompositeChange;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.engine.history.change.HistoryChange;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.apache.commons.text.StringSubstitutor;
+import org.triplea.java.collections.IntegerMap;
+
+/**
+ * Kills a set of units in a location
+ *
+ * <p>Adds a history event for the killing
+ *
+ * <p>Transforms units to other units if needed. See {@link TransformUnits}
+ */
+@Value
+public class KillUnits implements HistoryChange {
+
+  Territory location;
+  Collection<Unit> killedUnits;
+  Map<Territory, Collection<Unit>> unloadedUnits = new HashMap<>();
+  TransformUnits transformUnits;
+  String messageTemplate;
+  /** Units that were killed */
+  Collection<Unit> oldUnits = new ArrayList<>();
+  /** The units that were created after a transformation */
+  Collection<Unit> newUnits = new ArrayList<>();
+
+  public KillUnits(final Territory location, final Collection<Unit> killedUnits) {
+    this(location, killedUnits, "${units} lost in ${territory}");
+  }
+
+  /** @param messageTemplate ${units} and ${territory} will be replaced in this template */
+  public KillUnits(
+      final Territory location, final Collection<Unit> killedUnits, final String messageTemplate) {
+    this.location = location;
+    this.messageTemplate = messageTemplate;
+
+    // temporarily give the unit maximum damage so that TransformUnits will be able to
+    // recognize units that need to be transformed when they are killed
+    final IntegerMap<Unit> originalHits = new IntegerMap<>();
+    killedUnits.forEach(
+        unit -> {
+          originalHits.add(unit, unit.getHits());
+          unit.setHits(unit.getUnitAttachment().getHitPoints());
+        });
+
+    transformUnits = new TransformUnits(location, killedUnits);
+
+    killedUnits.forEach(
+        unit -> {
+          unit.setHits(originalHits.getInt(unit));
+        });
+
+    oldUnits.addAll(killedUnits);
+    // ensure that any units that are being transported are also killed
```

Comment:
Does this assume that transported units are always killed? Could we instead make this logic the responsibility of the caller?

Knowing that transported units are killed is a game-rule detail. This history object therefor not only tracks the history transformation, but it also encodes the rule that transported units are killed. If the caller were to handle this detail, and pass the list of transported units into the killed list, then this class would no longer be aware of that rule detail.

Reply:
Yes, it assumes that transported units are always killed.  There is currently no way to prevent that from happening in the engine.

I don't want the caller to handle this detail.  Because currently, it handles it incorrectly.  Some parts of the battle code that kills units will correctly remove their transported units.  Other parts of the battle code leaves the units floating in the ocean and luckily some other part of the code much later happens to clear them up.  By hiding this detail from the caller, it makes the code easier to read.

As for allowing the map to control whether this should be killed or not, that is the work I'm doing with the translate object code where transported items can be copied from one unit to another unit.